### ISSUE TITLE
Performance issue in Firefox with ng-container inside a <select>

### ISSUE THREAD

**Issue by laurentgoudet** [angular/angular#50233]
### Which @angular/* package(s) are the source of the bug?

platform-browser

### Is this a regression?

No

### Description

I have a list of ~2000 entries (US Universities), that I'd like to put in a `<select>` element, with some dynamic filtering done with an `ngIf` - in short the code looks like this:

```
    <select>
      <ng-container *ngFor="let option of options">
        <option
          *ngIf="true"
          [ngValue]="option.value"
        >
          {{ option.displayText ?? 'Please select' }}
        </option>
      </ng-container>
    </select>
```

While this works fine in Chrome, this causes Firefox to freeze. Spinning up the Firefox DevTools shows that the native DOM `Node.insertBefore` for each of the 2000 select items take almost 1s, compare to ~300ms in Chrome (x3000 faster).

Chrome:

<img width="1624" alt="Screenshot 2023-05-05 at 09 48 12" src="https://github.com/angular/angular/assets/2169585/9e817714-25e8-481f-a9de-f47ce39e7496">

Firefox:

<img width="1624" alt="Screenshot 2023-05-05 at 09 49 05" src="https://github.com/angular/angular/assets/2169585/8cfa6a87-1889-4f90-8e54-20c2cac52a90">

Playing around with the repro in https://stackblitz.com/edit/angular-f1vhwu?file=src%2Fmain.ts shows that removing the `*ngIf` & moving the `*ngFor` to the `<option>` element works around the issue, although in my case I cannot do that (i.e. pre-filter the `options` object) as I use the `*ngIf` as a type cast).

In addition, this seems to be specific to `<ng-container>` inside `<select>` element, as a simple list of `<div>`s load fine in Firefox (https://stackblitz.com/edit/angular-akq8zj?file=src/main.ts):

```
    <ng-container *ngFor="let option of options">
      <div *ngIf="true">
        {{ option.displayText ?? 'Please select' }}
      </div>
    </ng-container>
```

In short there seem to be a performance bug specific to Firefox with the way Angular handles `<ng-container>` elements within `<select>`

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/edit/angular-f1vhwu?file=src%2Fmain.ts

### Please provide the exception or error you saw

```true
Firefox becomes unresponsive and shows the "This page is slowing down Firefox.." banner
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
I'm on v14 but I've upgraded the StackBlitz to latest and it still happens
```

### Anything else?

Our CEO happens to be using Firefox 😀

**Comment by JeanMeche** [comment]
Hi, did you try using [trackBy](https://angular.io/api/core/TrackByFunction#description) to improve global performances when using an `*ngFor` ?

**Comment by laurentgoudet** [comment]
> In short, there seem to be a performance bug specific to Firefox with the way Angular handles <ng-container> elements within <select>

Actually that's only true when there's an `*ngIf` on the `<option>` element, so it seems to be about having 2 nested structure directives within a `<select>` basically.

> Hi, did you try using [trackBy](https://angular.io/api/core/TrackByFunction#description) to improve global performances when using an `*ngFor` ?

I did & I do use `trackBy` in the production code but `trackBy` only allows DOM elements to be reused (moved) when the model is updated, whereas here in the repro `options` is only set once, so I left it out to simplify the code at the maximum.

**Comment by pkozlowski-opensource** [comment]
@laurentgoudet would moving `ngFor` to a `<ng-template>` element help? That is, doing `<ng-template ngFor let-option [ngForOf]="options">` instead of using `<ng-container>`?

https://stackblitz.com/edit/angular-txd7lf?file=src%2Fmain.ts

**Comment by laurentgoudet** [comment]
> @laurentgoudet would moving `ngFor` to a `<ng-template>` element help? That is, doing `<ng-template ngFor let-option [ngForOf]="options">` instead of using `<ng-container>`? It would reduce the number of comment nodes created in the DOM and _might_ help avoiding this specific slow path in Firefox.
> 
> https://stackblitz.com/edit/angular-txd7lf?file=src%2Fmain.ts

Nope, same result - that's just syntactic sugar right, i.e. there's no real difference?

<img width="1624" alt="Screenshot 2023-05-10 at 18 25 10" src="https://github.com/angular/angular/assets/2169585/e0b5f24c-7efa-4131-b758-13441fd7b575">

By the way I am testing on Firefox 113.0 (latest stable) but this happens on 109.0.1 or 112.0.2 as well for instance, i.e. it does not seem like to be a new issue. I had tested with Angular 14 in the StackBlitz as well, same thing.

Note that depending on your machine it may finally load after a while (> 10 seconds for me on an M1 Pro), but it is still infinitely slower than in Chrome. It's not an infinite loop or really broken basically, it's just that that specific `insertBefore` takes forever for some reason.

**Comment by AndrewKushnir** [comment]
@laurentgoudet this issue reminds me of #41330. When an `[ngValue]` is used, an internal logic of the Forms module performs extra operations to compare options in select element (the comparison logic needs a performance improvement to be O(N) vs O(N^2)). Could you please try to use `[value]="option.value"` on the `<option>` element and see if that helps?

**Comment by laurentgoudet** [comment]
> @laurentgoudet this issue reminds me of #41330. When an `[ngValue]` is used, an internal logic of the Forms module performs extra operations to compare options in select element (the comparison logic needs a performance improvement to be O(N) vs O(N^2)). Could you please try to use `[value]="option.value"` on the `<option>` element and see if that helps?

Hm no using `[value]` does not seem to make a difference: https://stackblitz.com/edit/angular-lqqrfw?file=src%2Fmain.ts.

I've played a bit with the Firefox DevTools with the StackBlitz and basically 100% of the time is spent in `nsINode::GetChildAt_Deprecated`:

<img width="1512" alt="Screenshot 2023-05-11 at 16 45 40" src="https://github.com/angular/angular/assets/2169585/6f4b8afb-72cf-466e-b003-73bb7df5a766">

Which seems to be defined there: https://hg.mozilla.org/mozilla-central/file/tip/dom/base/nsINode.cpp#l1772

```
nsIContent* nsINode::GetChildAt_Deprecated(uint32_t aIndex) const {
  if (aIndex >= GetChildCount()) {
    return nullptr;
  }

  nsIContent* child = mFirstChild;
  while (aIndex--) {
    child = child->GetNextSibling();
  }

  return child;
}
```

Nothing really fancy here, although this has linear time complexity. This does not explain why this works fine without the `*ngIf` though. I am not familiar with how structural directives are implemented in Angular but, could it has something to do with the `Comment` Nodes that those create? It does feel like as if it causes this algorithm to turn quadratic somehow
